c_src/scenic/ops/cairo_script_ops.c

```diff
@@ -0,0 +1,663 @@
+
+void do_fill_stroke(cairo_t* cr, bool fill, bool stroke)
+{
+  if (fill && stroke)
+    cairo_fill_preserve(cr);
+  else if (fill)
+    cairo_fill(cr);
+
+  if (stroke) cairo_stroke(cr);
+}
+
+void script_ops_draw_line(void* v_ctx,
+                          coordinates_t a,
+                          coordinates_t b,
+                          bool stroke)
+{
+  if (g_opts.debug_mode) {
+    log_script_ops_draw_line(log_prefix, __func__, log_level_info,
+                             a, b, stroke);
+  }
+
+  CAIROcontext_t* p_ctx = (CAIROcontext_t*)v_ctx;
+  cairo_move_to(p_ctx->cr, a.x, a.y);
+  cairo_line_to(p_ctx->cr, b.x, b.y);
+
+  if (stroke) cairo_stroke(p_ctx->cr);
+}
+
+void script_ops_draw_triangle(void* v_ctx,
+                              coordinates_t a,
+                              coordinates_t b,
+                              coordinates_t c,
+                              bool fill, bool stroke)
+{
+  if (g_opts.debug_mode) {
+    log_script_ops_draw_triangle(log_prefix, __func__, log_level_info,
+                                 a, b, c, fill, stroke);
+  }
+
+  CAIROcontext_t* p_ctx = (CAIROcontext_t*)v_ctx;
+  cairo_move_to(p_ctx->cr, a.x, a.y);
+  cairo_line_to(p_ctx->cr, b.x, b.y);
+  cairo_line_to(p_ctx->cr, c.x, c.y);
+
+  do_fill_stroke(p_ctx->cr, fill, stroke);
+}
+
+void script_ops_draw_quad(void* v_ctx,
+                          coordinates_t a,
+                          coordinates_t b,
+                          coordinates_t c,
+                          coordinates_t d,
+                          bool fill, bool stroke)
+{
+  if (g_opts.debug_mode) {
+    log_script_ops_draw_quad(log_prefix, __func__, log_level_info,
+                             a, b, c, d, fill, stroke);
+  }
+
+  CAIROcontext_t* p_ctx = (CAIROcontext_t*)v_ctx;
+  cairo_move_to(p_ctx->cr, a.x, a.y);
+  cairo_line_to(p_ctx->cr, b.x, b.y);
+  cairo_line_to(p_ctx->cr, c.x, c.y);
+  cairo_line_to(p_ctx->cr, d.x, d.y);
+
+  do_fill_stroke(p_ctx->cr, fill, stroke);
+}
+
+void script_ops_draw_rect(void* v_ctx,
+                          float w,
+                          float h,
+                          bool fill, bool stroke)
+{
+  if (g_opts.debug_mode) {
+    log_script_ops_draw_rect(log_prefix, __func__, log_level_info,
+                             w, h, fill, stroke);
+  }
+
+  CAIROcontext_t* p_ctx = (CAIROcontext_t*)v_ctx;
+
+  cairo_rectangle(p_ctx->cr, 0, 0, w, h);
+
+  do_fill_stroke(p_ctx->cr, fill, stroke);
+}
+
+void script_ops_draw_rrect(void* v_ctx,
+                           float w,
+                           float h,
+                           float radius,
+                           bool fill, bool stroke)
+{
+  if (g_opts.debug_mode) {
+    log_script_ops_draw_rrect(log_prefix, __func__, log_level_info,
+                              w, h, radius, fill, stroke);
+  }
+
+  CAIROcontext_t* p_ctx = (CAIROcontext_t*)v_ctx;
+  cairo_arc(p_ctx->cr, 0 + radius, 0 + radius, radius, 2 * (M_PI/2), 3 * (M_PI/2));
```

Comment:
@ringlej Can you double check if cairo lib doesn't have a PI constant. Usually w/ graphics libs, they provide their own PI constant which can differ in size/value from the one found in `math.h`

Reply:
https://cgit.freedesktop.org/cairo/tree/boilerplate/cairo-boilerplate.h?h=1.16.0#n94

Reply:
The cairo tutorial source code using M_PI from math.h